src/userfunctions.jl

```diff
@@ -13,11 +13,23 @@ Return the current residual vector of the problem
 function getresidual end
 
 """
-    getjacobian(problem)
+    getsystemmatrix(problem,solver)
```

Comment:
I'm a bit wary of "opening" this to the user, as I feel it complicates the usage a bit. 
What we can do, is to have this function internal (for use in e.g. `solve_nonlinear` defined as 
```julia 
getsystemmatrix(problem, FerriteSolver{<:NewtonSolver}) = getjacobian(problem)
getsystemmatrix(problem, FerriteSolver{<:SteepestDescent}) = getpreconditioner(problem)
```
What do you think about that?

Reply:
I had it before with `getpreconditioner`. I discarded it, because it is ambiguous w.r.t. the future where we maybe want to have preconditioners for the nonlinear solve itself (stuff like this: https://www.researchgate.net/publication/362742563_Adaptive_Nonlinear_Domain_Decomposition_Methods_with_an_Application_to_the_p-Laplacian) I think the `getsystemmatrix` should be exposed to the user, since it seems natural to me that I use different matrices for different solvers and if I don't I can simply take the union or make a generic dispatch

`getsystemmatrix(problem, FerriteSolver) = ...`

Reply:
That's where I took the name from:) 
How would this work for dynamic problems, i.e. when we want both mass and stiffness matrices?
Or, for a solver that adaptively switches between the steepest descent and standard Newton?

I see the point that naming gets hard though, perhaps `getdescentpreconditioner` could work in this specific case?

For me, I think it is natural that a solvers library tells the user what it needs, not that the user needs to dispatch on the internal parts of the solver.

Reply:
I think a feasible strategy for dynamic problems is - or in general - to have a function which returns a time independent matrix, so something like `getsystemmatrix` and another function that provides the mass matrix, since they should be definitely separated in DAE systems. As far as I know, OrdinaryDiffEq.jl does exactly this splitting by taking a Jacobian and a mass matrix as separate arguments. They subdivided even further by having a linear part, a nonlinear part and the mass matrix.

For the solver switch, I imagine a nonlinear solver which has a `SteepestDescent` as well as a `Newton` as its fields, maybe in a vector (detailed data structure doesn't matter, but it just wraps those two in a certain way). This solver checks for convergence and for "normal convergence" you pass the `Newton` solver struct as a second argument whereas for failed convergence you pass the `SteepestDescent` solver to the function call. 

If you do 
```julia
getsystemmatrix(problem, ferritesolver) = problem.K
```
then it's not really dispatching on internals of the package, but I see in general what you mean.